contracts/Marketplace.sol

```diff
@@ -413,4 +418,66 @@ contract Marketplace is Ownable, IMarketplace {
     function reInitialize(address datacoinAddress, address currencyUpdateAgentAddress, address prev_marketplace_address) public onlyOwner {
         _initialize(datacoinAddress, currencyUpdateAgentAddress, prev_marketplace_address);
     }
+
+    //whitelist functionality
+
+    //should the whitelist getter be built into getProduct()?
+    function getRequiresWhitelist(bytes32 productId) public view returns (bool) {
+        (,address _owner,,,,,) = getProduct(productId);
+        require(_owner != 0x0, "error_notFound");
+        //if it's not local this will return 0, which is false
+        Product storage p = products[productId];
+        return p.requiresWhitelist;
+    }
+
+    function setRequiresWhitelist(bytes32 productId, bool _requiresWhitelist) public onlyProductOwner(productId) {
+        _importProductIfNeeded(productId);
+        Product storage p = products[productId];
+        require(p.id != 0x0, "Product not found");
+        p.requiresWhitelist = _requiresWhitelist;
+        if(_requiresWhitelist)
+            emit WhitelistEnabled(productId);
+        else
+            emit WhitelistDisabled(productId);
+    }
+
+    function whitelistApprove(bytes32 productId, address subscriber) public onlyProductOwner(productId) {
+        _importProductIfNeeded(productId);
+        Product storage p = products[productId];
+        require(p.id != 0x0, "Product not found");
+        require(p.requiresWhitelist, "Whitelist not enabled");
+        p.whitelist[subscriber] = WhitelistState.Approved;
+        emit WhitelistApproved(productId,subscriber);
+    }
+
+    function whitelistReject(bytes32 productId, address subscriber) public onlyProductOwner(productId) {
+        _importProductIfNeeded(productId);
+        Product storage p = products[productId];
+        require(p.id != 0x0, "Product not found");
+        require(p.requiresWhitelist, "Whitelist not enabled");
+        p.whitelist[subscriber] = WhitelistState.Rejected;
+        emit WhitelistRejected(productId,subscriber);
+    }
+
+    function whitelistRequest(bytes32 productId) public {
+        _importProductIfNeeded(productId);
+        Product storage p = products[productId];
+        require(p.id != 0x0, "Product not found");
+        require(p.requiresWhitelist, "Whitelist not enabled");
+        require(p.whitelist[msg.sender] == WhitelistState.None, "Whitelist request already submitted");
+        p.whitelist[msg.sender] = WhitelistState.Pending;
+        emit WhitelistRequested(productId,msg.sender);
+    }
+
+    // Do we need a getter like this?
+    /*
+    function getWhitelistState(bytes32 productId, address subscriber) public view returns (WhitelistState wlstate) {
+        (,address _owner,,,,,) = getProduct(productId);
```

Comment:
Do we need a getter like this? This contract has reached the deployment gas limit (~6700000) for the truffle test env, and any added bit of code increases deployment cost.

Reply:
I guess this was resolved by turning on optimizations

Reply:
> I guess this was resolved by turning on optimizations

Yes, with optimization enabled in package.json, 6000000 gas is again sufficient.